cpp/src/arrow/util/logging.h

```diff
@@ -202,48 +169,81 @@ class ARROW_EXPORT ArrowLog : public ArrowLogBase {
  private:
   ARROW_DISALLOW_COPY_AND_ASSIGN(ArrowLog);
 
+  std::ostream* Stream() override;
+
   // Hide the implementation of log provider by void *.
   // Otherwise, lib user may define the same macro to use the correct header file.
   void* logging_provider_;
   /// True if log messages should be logged and false if they should be ignored.
   bool is_enabled_;
 
   static ArrowLogLevel severity_threshold_;
-
- protected:
-  std::ostream& Stream() override;
 };
 
-// This class make ARROW_CHECK compilation pass to change the << operator to void.
-// This class is copied from glog.
-class ARROW_EXPORT Voidify {
- public:
-  Voidify() {}
-  // This has to be an operator with a precedence lower than << but
-  // higher than ?:
-  void operator&(ArrowLogBase&) {}
-};
+struct ARROW_EXPORT InterceptComparison {
+  template <typename NotBinaryOrNotPrintable>
+  static ArrowLogBase&& PrintOperands(const NotBinaryOrNotPrintable&,
+                                      ArrowLogBase&& log) {
+    return std::move(log);
+  }
 
-namespace detail {
+  template <typename Lhs, typename Rhs>
+  struct BoundLhsRhs {
+    explicit constexpr operator bool() const { return false; }
```

Comment:
Hmm... why?

Reply:
These bound expressions must be contextually convertible to bool to enable `DCHECK(x >= 0 && x < 8)`. (It can't be destructured for better printing, but it should still be a valid assertion.) If the conversion to bool is not provided, then `bound_expr && x < 8` breaks compilation

Reply:
Would there be a less cryptic way to implement the whole thing?